src/compiler/transformers/esnext.ts

```diff
@@ -302,9 +303,9 @@ export function transformESNext(context: TransformationContext): (x: SourceFile
             // before handing the shallow transformation back to the visitor for an in-depth transformation.
             const forInitializer = node.initializer;
             Debug.assertNode(forInitializer, isUsingVariableDeclarationList);
-            Debug.assert(forInitializer.declarations.length === 1, "ForInitializer may only have one declaration");
+            Debug.assert(forInitializer.declarations.length <= 1, "ForInitializer may only have one declaration");

```

Comment:
That was a crash on those broken empty declaration lists produced by the added test cases.

Other broken inputs are somewhat handled by transformers. For example:
```ts
// input

for (const x of)

// es5 output

for (var _i = 0, _1 = ; _i < _1.length; _i++) {
    var x = _1[_i];
    ;
}
```

It's not that this particular output is problem-free... it's broken but at least the transform/emit doesn't crash.